### PR TITLE
Add DataflowStartYamlJobOperator

### PR DESCRIPTION
**Pull request by e-galan**
Add DataflowStartYamlJobOperator as a replacement option for DataflowStartSqlJobOperator.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 1]
```diff
@@ -938,6 +937,182 @@ def launch_job_with_flex_template(
...
+
+        if on_new_job_callback:
+            on_new_job_callback(job)
+
+        jobs_controller.wait_for_done()
+
+        return jobs_controller.fetch_job_by_id(job_id=job_id)
+
+    @GoogleBaseHook.fallback_to_default_project_id
+    def launch_beam_yaml_job_deferrable(
```
Comment (moiseenkov) [inline airflow/providers/google/cloud/hooks/dataflow.py:1002]: I'd refactor the hook in the way that we had only one method `launch_beam_yaml_job()` which starts the job and returns it without waiting, and the operator would handle waiting for it depending on usage of deferrable mode. I think I could do it, once I'm free from higher priority tickets.

**Review by moiseenkov — CHANGES_REQUESTED** [review]

**Review by moiseenkov — APPROVED** [review]

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 2]
```diff
@@ -1050,6 +1058,177 @@ def on_kill(self) -> None:
...
+                    expected_terminal_state=self.expected_terminal_state,
+                    impersonation_chain=self.impersonation_chain,
+                ),
+                method_name=GOOGLE_DEFAULT_DEFERRABLE_METHOD_NAME,
+            )
+
+        self.hook.wait_for_done(
+            job_name=self.job_name, location=self.region, project_id=self.project_id, job_id=self.job["id"]
+        )
+        return self.job
```
Comment (e-galan) [inline airflow/providers/google/cloud/operators/dataflow.py:1192]: Hey, I don't mean to intrude (still getting emal notifications about this PR), but I think in this way you would return the job data in an outdated state in the sync mode. You'll need to add one more request to fetch / get the job data before returning the results. To explain:

 1) On line 1162 you're getting data from a job that has just been created and is probably still in the QUEUED state.
 2) On line 1189 you're waiting for the job to complete, but you don't update the `self.job` attribute. Which means that 
 3) `return self.job` will return the data in the same state it was on line 1162

Comment (moiseenkov) [reply airflow/providers/google/cloud/operators/dataflow.py]: Thanks for the suggestion. I covered this.

**Comment by moiseenkov** [comment]
@MaksYermak , hi
Please take a look

File: airflow/providers/google/cloud/hooks/dataflow.py [hunk 3]
```diff
@@ -938,6 +937,121 @@ def launch_job_with_flex_template(
...
+        cmd = self._build_yaml_gcloud_command(
+            job_name=job_name,
+            yaml_pipeline_file=yaml_pipeline_file,
+            project_id=project_id,
+            region=location,
+            options=options,
+            jinja_variables=jinja_variables,
+        )
+        job_id = self._create_dataflow_job_with_gcloud(cmd=cmd)
+        _DataflowJobsController(
```
[the remark this answers is not in the document]
Comment (moiseenkov) [reply airflow/providers/google/cloud/hooks/dataflow.py]: Yes, you're right! Thank you

**Review by MaksYermak — APPROVED** [review]

**Pull request closed without merging**